internal/remotestate/backend/gcs/client.go

```diff
@@ -0,0 +1,399 @@
+package gcs
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"time"
+
+	"cloud.google.com/go/storage"
+	"github.com/gruntwork-io/terragrunt/internal/errors"
+	"github.com/gruntwork-io/terragrunt/internal/remotestate/backend"
+	"github.com/gruntwork-io/terragrunt/options"
+	"github.com/gruntwork-io/terragrunt/pkg/log"
+	"github.com/gruntwork-io/terragrunt/shell"
+	"github.com/gruntwork-io/terragrunt/util"
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/jwt"
+	"google.golang.org/api/impersonate"
+	"google.golang.org/api/iterator"
+	"google.golang.org/api/option"
+)
+
+const (
+	maxRetriesWaitingForGcsBucket          = 12
+	sleepBetweenRetriesWaitingForGcsBucket = 5 * time.Second
+
+	gcpMaxRetries          = 3
+	gcpSleepBetweenRetries = 10 * time.Second
+)
+
+type Client struct {
+	*ExtendedRemoteStateConfigGCS
+	*storage.Client
+	logger log.Logger
+}
+
+// NewClient inits GCS client.
+func NewClient(ctx context.Context, config *ExtendedRemoteStateConfigGCS, logger log.Logger) (*Client, error) {
+	var opts []option.ClientOption
+
+	gcsConfig := config.RemoteStateConfigGCS
+
+	if gcsConfig.Credentials != "" {
```

Comment:
NIT: I don't think this is new code, but it's pretty confusing.

It's hard to understand what's happening in each branch here. It would be easier to understand with a switch and/or well named functions.

Reply:
Agreed, it's ugly, but not my code. Let's do any further improvement in another PR.